### ISSUE TITLE
Issue reproducing paper results

### ISSUE THREAD

**Issue by Kuan2vn** [DZhaoXd/SND#8]
Hello, I trained on the full GTA5 dataset using `train_src.py`, and then performed adaptation to Cityscapes with `train_SND.py`. However, my current result is only around 0.51 mIoU, which is significantly lower than the performance reported in the paper. Could you please share the exact training settings or configurations needed to reproduce the reported results? Thank you!

**Comment by DZhaoXd** [comment]
Thanks for your interest. I think the main problem may be that the adaptation was not initialized from the proper pretrained source model. In this repo, the intended workflow is to first load the provided source-only pretrained checkpoint (for example the GTA5 pretrained model), check that it already gives a reasonable starting result on Cityscapes, and then perform SND adaptation from that checkpoint.

**Comment by Kuan2vn** [comment]
Thanks for your clarification.

As I mentioned in the previous issue, the link to the pretrained source model checkpoint appears to be no longer available. I noticed that you provided train_src.py, which suggests training the source model from scratch as an alternative.

However, based on your explanation, it seems that using the original pretrained checkpoint is important for achieving the reported results. Would it be possible for you to re-upload or share the pretrained source model checkpoint?

Thank you very much for your help!

**Comment by DZhaoXd** [comment]
I have checked the pretrained checkpoints in the google drive, but unfortunaly, it was deleted due to the limited space. I think you can use the train_src.py to reproduce the pretrain results.

**Comment by Kuan2vn** [comment]
Thanks for your response.

I have trained the source model using `train_src.py` with the configuration from `deeplabv2_r101_dtst_G2C.yaml`. However, after 20k iterations, the validation mIoU is only **0.4077**.

Then, I used this checkpoint to run `train_SND.py` with `deeplabv2_r101_SND_G2C.yaml`, but the final result is only around **0.51 mIoU**.

Could you please share more detailed configuration settings (e.g., training iterations, DTU, learning rate schedule, data augmentation, or any important tricks) that you used to obtain the reported results?

Thank you very much for your support!

**Comment by DZhaoXd** [comment]
Hi could you add my wechat for detailed discussion?  zd1564421854

**Comment by Kuan2vn** [comment]
Hi, thanks for reaching out. I don’t have WeChat—would it be possible to discuss via Telegram instead? You can contact me at @mkhoaz

**Comment by DZhaoXd** [comment]
I don’t have Telegram. I try to regist one account. 

From your description, one possible reason is that the warm-up stage may not have been completed before running the full SND training. In our pipeline, SND relies on the stored stable neighbor samples produced during the warm-up stage. If these neighbor samples are not properly generated or saved, the subsequent SND training may degrade to a much weaker self-training process, which could explain the lower final mIoU.

Could you please first check the following points?

Please make sure you have run the warm-up stage before the full SND training, for example:
python train_SND.py \
  -cfg configs/deeplabv2_r101_dtst_G2C.yaml \
  OUTPUT_DIR results/G2C_SND_WARM/ \
  resume pretrain/G2C_model_iter020000.pth

After the warm-up stage, please check whether the stable neighbor samples have been correctly stored in the output directory. The full SND training should be able to load these saved neighbor samples.
Please also check the visualization results generated during the warm-up / neighbor selection stage. If the selected neighbors or pseudo-label visualizations look abnormal, it usually indicates that the neighbor memory was not built correctly.

Then run the full SND training with:
python train_SND.py \
  -cfg configs/deeplabv2_r101_SND_G2C.yaml \
  OUTPUT_DIR results/G2C_SND/ \
  resume pretrain/G2C_model_iter020000.pth

In our experiments, this warm-up stage is important because it provides reliable stable neighbors for denoising. Without it, the final performance can be much lower than the reported results.

Please let me know whether the stable neighbor files and the corresponding visualizations are correctly generated after the warm-up stage. Then we can further check whether the issue comes from the source model, the data split, or the SND training configuration.

**Comment by Kuan2vn** [comment]
It seems I indeed missed the warm-up step. Thanks a lot for pointing that out — I’ll run the experiment again.

**Comment by Kuan2vn** [comment]
Quick update: I retrained with the warm-up step, but the result only improves to around 0.53 mIoU.

I checked the stable neighbor outputs after 7000 warm-up iterations — they are still somewhat noisy, but already show reasonable structure. I then copied all the `CTR` and `meta_val` files/folders into the output directory for full SND and ran the full SND stage as you suggested.

The training converges at about 0.53 mIoU after ~7500 iterations and then starts to decrease.

Do you have any suggestions on what might be going wrong or what I should further check?

**Comment by DZhaoXd** [comment]
In short, I think the current result may mainly come from a gap in the pre-trained source model. SND depends heavily on the source model quality, so a weaker source model can lead to noisy stable-neighbor estimation and lower final adaptation performance.

**Comment by Kuan2vn** [comment]
Thanks for the clarification — that makes sense.

Given that the source model quality is critical, could you please suggest more specific configurations (e.g., training iterations, learning rate, batch size, augmentations, etc.) to obtain a stronger source model?

Alternatively, is there any place where I could download a suitable pretrained checkpoint that matches your setup?

Thank you again for your help!

**Comment by DZhaoXd** [comment]
you can try the pretrain model: https://drive.google.com/file/d/1OcQ9EkaSCrez7lh2kPBVahevDdacFFq2/view?usp=drive_link
